internal/server/server.go

```diff
@@ -17,209 +17,255 @@ import (
+	tools := openapi2mcp.RegisterOpenAPITools(srv, operations, oasDoc, nil)
 
-	return srv, nil
+	return srv, tools, nil
 }
 
-func startHTTPServer(
-	ctx context.Context,
-	mcpSrv *mcpserver.MCPServer,
-	listenAddr string,
-	serveOpts *ServeOptions,
-) *CustomStreamableHTTPServer {
-	// Wrapper to pass the url and other params in the future
-	authContextFuncWrapper := func(c context.Context, r *http.Request) context.Context {
-		if !serveOpts.OauthEnabled {
-			return authContextFuncNoOauth(
-				c,
-				r,
-				serveOpts.OauthValidateURL,
-				serveOpts.TrentoURL,
-				serveOpts.TrentoUsername,
-				serveOpts.TrentoPassword,
-			)
-		}
+// handleServerRun configures and starts the appropriate server based on the selected transport.
+// It sets up an authentication context wrapper and blocks until a shutdown signal is received.
+func handleServerRun(ctx context.Context, srv *mcpserver.MCPServer, serveOpts *ServeOptions) error {
+	// Build the address to listen to
+	listenAddr := fmt.Sprintf(":%d", serveOpts.Port)
+
+	slog.DebugContext(ctx, "about to start the MCP server",
+		"server.address", listenAddr,
+		"server.transport", serveOpts.Transport,
+	)
 
+	// Wrapper to pass the url and other params in the future
+	authContext := func(c context.Context, r *http.Request) context.Context {
 		return authContextFunc(
 			c,
 			r,
+			serveOpts.OauthEnabled,
 			serveOpts.OauthValidateURL,
 			serveOpts.TrentoURL,
 			serveOpts.TrentoUsername,
 			serveOpts.TrentoPassword,
 		)
 	}
 
-	// Create the server, using custom one to handle mcp auth
-	streamableServer := NewCustomStreamableHTTPServer(mcpSrv, "/mcp", authContextFuncWrapper, serveOpts)
+	serverErrChan := make(chan error, 1)
+	var stoppableServer StoppableServer
+	var err error
 
-	// Run the http server for the mcp in a separate goroutine.
-	go func() {
-		slog.InfoContext(ctx, "mcp server via HTTP starting", "server.address", listenAddr)
+	// Depending on the chosen transport, we handle the server startup.
+	switch serveOpts.Transport {
+	case utils.TransportSSE:
+		stoppableServer, err = startSSEServer(ctx, srv, listenAddr, serveOpts, authContext, serverErrChan)
 
-		err := streamableServer.Start(listenAddr)
-		if err != nil && !errors.Is(err, http.ErrServerClosed) {
-			slog.ErrorContext(ctx, "failed to serve MCP server via HTTP", "error", err)
-		}
-	}()
+	case utils.TransportStreamable:
+		stoppableServer, err = startStreamableHTTPServer(ctx, srv, listenAddr, serveOpts, authContext, serverErrChan)
 
-	slog.InfoContext(ctx, "mcp server via HTTP listening successfully", "server.address", listenAddr)
+	default:
+		return fmt.Errorf("invalid transport type: %s", serveOpts.Transport)
+	}
+	if err != nil {
+		return err
+	}
 
-	return streamableServer
+	// Block until shutdown or an error occurs
+	return waitForShutdown(ctx, stoppableServer, serverErrChan)
 }
 
-func startSSEServer(
+// startServer starts a given MCP in a goroutine.
+func startServer(
```

Comment:
I've tried to reduce code duplication for the `streamable` and `sse` branches. I think we need to support both for now. In the past, I had to use one of the transports (SSE, I'd say?) for the auth flow to work in the `mcpo` proxy.